[quant] Add reference quantized linear/linear_relu module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48192 [quant] Add reference quantized linear/linear_relu module**

Summary:
Some backend don't have packed weight for linear/linear_relua, this adds a reference
module that does not contain packed weight for linear/linear_relu

Test Plan:
test_quantized_module.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25061645](https://our.internmc.facebook.com/intern/diff/D25061645)

jerryzh168 committed:
[quant] Add backend_independent option for quantized linear module

Summary:
This is to allow producing a backend independent quantized module
since some backend don't have packed weight for linear

Test Plan:
test_quantized_module.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add backend_independent option for quantized linear module"


Summary:
This is to allow producing a backend independent quantized module
since some backend don't have packed weight for linear

Test Plan:
test_quantized_module.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25061645](https://our.internmc.facebook.com/intern/diff/D25061645)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add backend_independent option for quantized linear module"


Summary:
This is to allow producing a backend independent quantized module
since some backend don't have packed weight for linear

Test Plan:
test_quantized_module.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25061645](https://our.internmc.facebook.com/intern/diff/D25061645)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add backend_independent option for quantized linear module"


Summary:
This is to allow producing a backend independent quantized module
since some backend don't have packed weight for linear

Test Plan:
test_quantized_module.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25061645](https://our.internmc.facebook.com/intern/diff/D25061645)

[ghstack-poisoned]

raghuramank100 commented on torch/nn/intrinsic/quantized/modules/__init__.py:
```diff
@@ -1,10 +1,10 @@
 
 __all__ = [
     'LinearReLU',
+    'LinearReLUBackendIndependent',
```
NIT: Should we create a namespace nn.quantized.reference and have Linear, LinearRelu in there? The name BackendIndependent could be confusing to external users. Also, if we dont intend this for direct use, should we prefix with an _?, i.e have:
nn.quantized._reference.LinearRelu

jerryzh168 replied:
Sure

jerryzh168 replied:
@raghuramank100 do you also suggest to change backend_independent to reference in the api?

jerryzh168 replied:
changed to reference

raghuramank100 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -123,11 +124,12 @@ class Linear(torch.nn.Module):
         torch.Size([128, 30])
     """
-    _version = 3
+    _version = 4
```
Why is the version number bumped up here?

jerryzh168 replied:
sorry, this is left over from previous changes, will revert

jerryzh168 committed:
Update on "[quant] Add backend_independent option for quantized linear module"


Summary:
This is to allow producing a backend independent quantized module
since some backend don't have packed weight for linear

Test Plan:
test_quantized_module.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25061645](https://our.internmc.facebook.com/intern/diff/D25061645)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add backend_independent option for quantized linear module"


Summary:
This is to allow producing a backend independent quantized module
since some backend don't have packed weight for linear

Test Plan:
test_quantized_module.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25061645](https://our.internmc.facebook.com/intern/diff/D25061645)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add reference quantized linear/linear_relu module"


Summary:
Some backend don't have packed weight for linear/linear_relua, this adds a reference
module that does not contain packed weight for linear/linear_relu

Test Plan:
test_quantized_module.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25061645](https://our.internmc.facebook.com/intern/diff/D25061645)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add reference quantized linear/linear_relu module"


Summary:
Some backend don't have packed weight for linear/linear_relua, this adds a reference
module that does not contain packed weight for linear/linear_relu

Test Plan:
test_quantized_module.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25061645](https://our.internmc.facebook.com/intern/diff/D25061645)

[ghstack-poisoned]
